libs/pika/async_mpi/src/mpi_polling.cpp

```diff
@@ -688,168 +736,189 @@ namespace pika::mpi::experimental {
+                    "Register failed '{}' does not exist", pool_name);
+            PIKA_DETAIL_DP(
+                mpi_debug<1>, debug(str<>("register_pool"), "pool =", polling_pool_name_));
         }
-        // use reference so that we change the actual value in place
-        std::size_t& flags = detail::completion_flags_;
-        if (mode == pool_create_mode::force_create)
-            flags |= 1;
-        else if (mode == pool_create_mode::force_no_create)
-            flags &= ~1;
-        else if ((mode == pool_create_mode::pika_decides) && (detail::mpi_data_.size_ == 1))
+
+        // -------------------------------------------------------------
+        void unregister_polling()
         {
-            // if we have a single rank - disable pool
-            PIKA_DETAIL_DP(detail::mpi_debug<1>, debug(str<>("single rank"), "Pool disabled"));
-            flags &= ~1;
+            auto mode = get_completion_mode();
+            // enable pika polling on the mpi pool if the handling mode needs it
+            if (detail::get_handler_method(mode) != handler_method::yield_while)
+            {
+                PIKA_DETAIL_DP(detail::mpi_debug<1>,
+                    debug(str<>("disabling polling"), "pool =", get_pool_name(), ", mode",
+                        mode_string(get_completion_mode()), get_completion_mode()));
+                detail::unregister_polling(pika::resource::get_thread_pool(get_pool_name()));
+            }
         }
-        PIKA_DETAIL_DP(detail::mpi_debug<1>,
-            debug(str<>("completion mode"), bin<8>(flags), detail::mode_string(flags)));
-        // override the variable used to control completion mode and pool flags
-        setenv("PIKA_MPI_COMPLETION_MODE", std::to_string(flags).c_str(), true);
 
-        // if pool is now disabled, just exit
-        if (!detail::use_pool(flags))
+        // -----------------------------------------------------------------
+        bool create_pool(pika::resource::partitioner& rp, std::string const& pool_name,
+            polling_pool_creation_mode mode)
         {
-            set_pool_name("default");
-            return false;
+            // if we are potentially using a pool, mpi must be initialized
+            if (!mpi::detail::environment::is_mpi_initialized())
+            {
+                PIKA_THROW_EXCEPTION(pika::error::invalid_status, "mpi::create_pool",
+                    "MPI must be initialized prior to pool creation");
+            }
+            // needed, in order to decide if pool can be disabled on 1 rank
+            mpi_data_.rank_ = mpi::detail::environment::rank();
+            mpi_data_.size_ = mpi::detail::environment::size();
+
+            bool do_create = true;
+            if (mode == polling_pool_creation_mode::mode_pika_decides)
+            {
+                if (mpi_data_.size_ == 1)
+                {
+                    PIKA_LOG(warn, "MPI Pool creation disabled, running on a single rank");
```

Comment:
I'm undecided on this. Usually you develop a test or app and run it, then to debug something, you might try running it on one rank to see if a problem still exists - with all other options unchanged - now pika will turn off the pool (if you had been using it) and you might not have expected that - it might substantially change the scheduling of tasks etc compared to what you had been testing with previously - is that a warning or an info?

Reply:
Yeah, I agree it's a borderline case. My thinking is that it's expected behaviour that it gets disabled and there's not really anything the user can do at the moment (at runtime) to either force the creation of the pool or disable the warning (except by disabling _all_ warnings). It's (IMO) more "FYI" (i.e. `info`) rather than "you probably did something wrong, please check" (i.e. `warn`).

With a single rank there shouldn't be any (important) MPI work being done anyway.=

> it might substantially change the scheduling of tasks etc

How significant do you expect this to be? Not doing MPI calls on a single rank would already change scheduling of tasks. In the ideal case having or not having the MPI pool disabled with a single rank should mean you have one thread less, which should have minimal impact. (Keyword: should).

That said, I get your point and I'm not set on the above.

Most importantly: the behaviour should be documented on https://pikacpp.org/usage.html.

Reply:
Addition: logging currently prints warnings and worse by default. Info isn't printed by default.

Reply:
I don't think we have a command-line way of forcing the creation of the pool on 1 rank. :(

Reply:
Indeed, I don't think so either, but I also don't think it really makes sense to have that (except possibly as a debug/internal option; I would definitely not advertise that option if it existed)?

Reply:
Let's keep it as info for now then and see if there's any feedback once merged (likely none).